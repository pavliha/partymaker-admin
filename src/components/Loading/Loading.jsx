import React, { Component } from 'react'
import { object, number, string, bool, node } from 'prop-types'
import { CircularProgress } from '@material-ui/core'
import { withStyles } from '@material-ui/styles'
import classNames from 'classnames'

const styles = {
  root: {
    height: '100%',
    width: '100%',
    display: 'flex',
    alignItems: 'center',
    flexDirection: 'column',
    justifyContent: 'center',
  },
}

class Loading extends Component {
  state = {
    isVisible: false,
  }

  componentDidMount() {
    const { debounce } = this.props
    this.timer = setTimeout(() => this.setState({ isVisible: true }), debounce)
  }

  componentWillUnmount() {
    clearTimeout(this.timer)
  }

  render() {
    const { classes, size, color, loaded, children, className } = this.props
    const { isVisible } = this.state

    if (!isVisible) {
      return null
    }

    if (loaded) return children

    return (
      <div className={classNames(classes.root, { [className]: !!className })}>
        <CircularProgress color={color} className={classes.progress} size={size} />
      </div>
    )
  }
}

Loading.propTypes = {
  classes: object.isRequired,
  size: number,
  debounce: number,
  color: string,
  loaded: bool,
  children: node,
  className: string,
}
Loading.defaultProps = {
  size: 80,
  debounce: 700,
  color: 'primary',
  loaded: false,
  children: null,
  className: null,
}

export default withStyles(styles)(Loading)
