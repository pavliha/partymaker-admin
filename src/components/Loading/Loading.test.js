import React from 'react'
import Loading from './Loading'
import { CircularProgress } from '@material-ui/core'

describe('<Loading />', () => {

  it('should be invisible before timeout', () => {
    const tree = shallow(<Loading />).dive()
    tree.setState({ isVisible: false })
    expect(tree.find(CircularProgress).exists()).toBeFalsy()
  })

  it('should show progress after timeout', () => {
    const tree = shallow(<Loading />).dive()
    tree.setState({ isVisible: true })
    expect(tree.find(CircularProgress).exists()).toBeTruthy()
  })

  it('should show children when loaded', () => {
    const tree = shallow(<Loading loaded>child</Loading>).dive()
    tree.setState({ isVisible: true })
    expect(tree.text()).toContain('child')
  })

})
