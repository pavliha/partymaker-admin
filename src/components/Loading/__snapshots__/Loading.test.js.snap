// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Loading /> should render properly 1`] = `
<WithStyles(Loading)
  className={null}
  color="primary"
  debounce={700}
  loaded={false}
  size={80}
>
  <Loading
    className={null}
    classes={
      Object {
        "root": "Loading-root-1",
      }
    }
    color="primary"
    debounce={700}
    loaded={false}
    size={80}
  />
</WithStyles(Loading)>
`;
